scipy/sparse/csgraph/_flow.pyx

```diff
@@ -407,3 +431,200 @@ cdef ITYPE_t[:] _edmonds_karp(
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    levels: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``levels[v]`` is the level of ``v`` in 
+        the layered graph of input graph.
+    q : memoryview of length :math:`|E|`
+        Queue to be used in breadth first search. Passed to avoid repeated 
+        queue creation inside this function.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the layered graph creation was successful,
+        otherwise ``False``.
+
+    """
+    cdef ITYPE_t n_verts = edge_ptr.shape[0] - 1
+
+    cdef ITYPE_t cur, start, end, dst_vertex, e
+
+    q[0] = source
+    start = 0
+    end = 1
+    levels[source] = 0
+
+    while start != end:
+        cur = q[start]
+        start += 1
+        if cur == sink:
+            return 1
+        for e in range(edge_ptr[cur], edge_ptr[cur + 1]):
+            dst_vertex = heads[e]
+            if capacities[e] > 0 and levels[dst_vertex] == -1:
+                levels[dst_vertex] = levels[cur] + 1
+                q[end] = dst_vertex
+                end += 1
+    return 0
+
+cdef ITYPE_t _augment_paths(
+        const ITYPE_t[:] edge_ptr, # IN
+        const ITYPE_t current, # IN
+        const ITYPE_t sink, # IN
+        const ITYPE_t[:] levels, # IN
+        const ITYPE_t[:] heads, # IN
+        const ITYPE_t[:] rev_edge_ptr, # IN
+        ITYPE_t[:] capacities, # IN/OUT
+        ITYPE_t[:] progress, # IN
+        ITYPE_t[:] flows, # OUT
+        ITYPE_t flow # OUT
+    ) nogil:
+    """Computes blocking flow in layered graph using depth first search.
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex ``v``, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    current : int
+        The current node in the path.
+    sink : int
+        The sink vertex.
+    levels: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``levels[v]`` is the level of ``v`` in 
+        the layered graph of input graph.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    rev_edge_ptr : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``rev_edge_ptr[e]`` is the edge obtained by
+        reversing ``e``. In particular, ``rev_edge_ptr[rev_edge_ptr[e]] == e``.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    progress: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``progress[v]`` is the index of the next
+        edge to be visited from ``v``.
+    flows : memoryview of length :math:`|E|`
+        The residual graph with respect to a maximum flow.
+    flow : int
+        Current maximum flow in the path.
+
+    Returns
+    -------
+    result_flow : int
+        The blocking flow in the layered graph.
```

Comment:
If I'm reading this correctly, this method only produces _a_ flow; only at the final step of the loop in `_dinic` calling this function does the flow become blocking.

Reply:
Addressed.